src/main/scala/org/constellation/p2p/Cluster.scala

```diff
@@ -0,0 +1,425 @@
+      client = APIClient(adjustedHost, pm.httpPort)(ConstellationExecutionContext.apiClient, dao)
+
+      _ <- C.shift *> peerDiscovery(client) // mwadon: make sure it works as expected
+
+      _ <- Sync[F].delay(client.id = pm.id)
+
+      _ <- updatePeerInfo(PeerData(pm, client))
+    } yield ()
+  }
+
+  def pendingRegistration(ip: String, request: PeerRegistrationRequest): F[Unit] =
+    for {
+      pr <- PendingRegistration(ip, request).pure[F]
+      _ <- Logger[F].debug(s"Pending Registration request: $pr")
+      validExternalHost = validWithLoopbackGuard(request.host)
+      peers <- peers.get
+      hostAlreadyExists = peers.exists {
+        case (_, d) => d.client.hostName == request.host && d.client.apiPort == request.port
+      }
+      validHost = validExternalHost && !hostAlreadyExists
+      isSelfId = dao.id == request.id
+      badAttempt = isSelfId || !validHost
+
+      _ <- if (badAttempt) {
+        dao.metrics.incrementMetricAsync[F]("duplicatePeerAdditionAttempt")
+      } else {
+        val client = APIClient(request.host, request.port)(ConstellationExecutionContext.apiClient, dao)
+        val authSignRequest = PeerAuthSignRequest(Random.nextLong())
+        val req = client.postNonBlockingF[F, SingleHashSignature]("sign", authSignRequest)
+
+        req.flatTap { sig =>
+          if (sig.hashSignature.id != request.id) {
+            Logger[F].warn(s"Keys should be the same: ${sig.hashSignature.id} != ${request.id}") *>
+              dao.metrics.incrementMetricAsync[F]("peerKeyMismatch")
+          } else Sync[F].unit
+        }.flatTap { sig =>
+          if (!sig.valid) {
+            Logger[F].warn(s"Invalid peer signature $request $authSignRequest $sig") *>
+              dao.metrics.incrementMetricAsync[F]("invalidPeerRegistrationSignature")
+          } else Sync[F].unit
+        }.flatTap { sig =>
+          dao.metrics.incrementMetricAsync[F]("peerAddedFromRegistrationFlow") *>
+            Logger[F].debug(s"Valid peer signature $request $authSignRequest $sig")
+        }.flatMap { sig =>
+          val state = withMetric(
+            client.getNonBlockingF[F, NodeStateInfo]("state"),
+            "nodeState"
+          )
+
+          state.flatMap { s =>
+            val id = sig.hashSignature.id
+            val add = PeerMetadata(
+              request.host,
+              request.port,
+              id,
+              nodeState = s.nodeState,
+              auxAddresses = s.addresses,
+              resourceInfo = request.resourceInfo
+            )
+            client.id = id
+            val peerData = PeerData(add, client)
+            updatePeerInfo(peerData) *> C.shift *> peerDiscovery(client)
+          }.void
+
+        }.handleErrorWith { err =>
+          Logger[F].warn(s"Sign request to ${request.host}:${request.port} failed. $err") *>
+            dao.metrics.incrementMetricAsync[F]("peerSignatureRequestFailed")
+        }
+      }
+    } yield ()
```

Comment:
this method has 60 lines, I do not know how it is in scala, but I feel that it is too long

Reply:
I know but it's almost 1:1 rewritten as it was in original PeerManager

Reply:
ok